### PR TITLE
`spine-codestyle-checker` plugin extension fix

### PR DESCRIPTION
**Pull request by dmytro-grankin**
This PR contains the `spine-codestyle-checker` extension fix, which makes the plugin configuration more flexible. Before this fix, the parameters for the `spine-codestyle-checker`'s sub checkers was common.

For now every kind of code style check has the two same parameters:
- `threshold` is a number of bearable code violations that will not trigger a report.
- `reportType` is a verification report type, that can be either `error` or `warn`.

So plugin configuration may look like:
```
codestyleChecker {

    javadocLinkChecker {

    threshold = 0
    reportType = "warn"
  }
}
```

### CODE REVIEW

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ */
+package org.spine3.tools.codestyle;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A configuration class for the {@link CodestyleCheckerPlugin} nested plugins.
+ *
+ * @author Alexander Aleksandrov
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java]: Seems the author is wrong.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ */
+public class SubcheckerExtension {
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java]: `StepConfiguration`

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleExtension.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+package org.spine3.tools.codestyle;
+
+/**
+ * A configuration class for the {@link CodestyleCheckerPlugin}.
+ *
+ * <p>Serves as a parent extension for {@link SubcheckerExtension}s.
+ *
+ * @author Dmytro Grankin
+ */
+public class CodestyleExtension {
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleExtension.java]: `CodestylePluginConfiguration`

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 4]
```diff
@@ -32,17 +33,26 @@
...
     public void apply(final Project project) {
         project.getExtensions()
-               .create(CODESTYLE_CHECKER_EXTENSION_NAME, Extension.class);
+               .create(CODESTYLE_CHECKER_EXTENSION_NAME, CodestyleExtension.class);
 
         log().debug("Applying Spine Javadoc link checker plugin");
         new JavadocLinkCheckerPlugin().apply(project);
     }
 
+    public static SubcheckerExtension createSubcheckerExtension(
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: `createStepExtension`.

Also, this method must be documented, as it is a part of public API.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.tools.codestyle;
+
+/**
+ * A bearable (non-negative) threshold of code style violations,
+ * that will not trigger a code style verification report.
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: The description is unclear. Please rephrase, probably, in a few sentences.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 6]
```diff
@@ -32,17 +33,26 @@
  */
 public class CodestyleCheckerPlugin extends SpinePlugin {
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: Please describe the general flow for the plugin, i.e. that its processing is done by executing several checks one by one. And please have a code example.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java (hunk 2 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleExtension.java (hunk 3 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleExtension.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java (hunk 6 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java (hunk 4 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java (hunk 5 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java (hunk 1 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/SubcheckerExtension.java]: Done.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+package org.spine3.tools.codestyle;
+
+/**
+ * A bearable (non-negative) threshold of code style violations.
+ *
+ * <p>If a code style violations amount not exceed a threshold value,
+ * then a code style verification is considered passed:
+ * <ul>
+ * <li>violations should not be reported;</li>
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: Please format the HTML with spaces to improve readability.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ * <li>if report type is {@linkplain ReportType#ERROR error}, build should not be failed.</li>
+ * </ul>that violations should not be reported
+ *
+ * @author Dmytro Grankin
+ * @see ReportType
+ */
+public class Threshold {
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: This class looks `final`.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    private final int value;
+
+    /**
+     * Creates a new instance.
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: Please expand the documentation by mentioning its behaviour for the negative values.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A configuration class for the {@link CodestyleCheckerPlugin} nested plugins.
+ *
+ * @author Dmytro Grankin
+ */
+public class StepConfiguration {
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java]: This class looks `final`.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+    private Threshold threshold = new Threshold(0);
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java]: Please describe the meaning of each of the parameters and their default values.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        return threshold;
+    }
+
+    public void setThreshold(int value) {
+        this.threshold = new Threshold(value);
+        log().debug("Threshold set up to {}", threshold);
+    }
+
+    public void setReportType(String reportType) {
+        this.reportType = ReportType.of(reportType);
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java]: I think we should be checking for `null`.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 13]
```diff
@@ -28,21 +29,62 @@
 /**
  * The plugin that verifies code style.
  *
+ * <p>The plugin verifies the few code style aspects.
+ *
+ * <p>Every verification starts one by one during a build process and
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]:
```
The verification consists of several steps, executed in a chain during the build process.

<p>Each step is configured with the parameters as follows:
<ul>
     <li> <here you describe the parameters in short> </li>
...
</ul>

<p> Example:
...
   <in the example you should actually explain what has been configured>
```

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 14]
```diff
@@ -28,21 +29,62 @@
...
+ *
+ *      anotherCodestyleChecker {
+ *
+ *          threshold = 5
+ *          reportType = "error"
+ *      }
+ * }
+ * }</pre>
+ *
  * @author Alexander Aleksandrov
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: Please append yourself as an author by adding another `@author` tag.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java (hunk 7 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java (hunk 8 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java (hunk 9 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/Threshold.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java (hunk 10 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java (hunk 11 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java (hunk 12 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/StepConfiguration.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java (hunk 13 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: Done.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java (hunk 14 above)
Comment (dmytro-grankin) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: Done.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 15]
```diff
@@ -28,21 +29,73 @@
...
+ *     <li>{@linkplain ReportType "reportType"} if a check is not passed.</li>
+ * </ul>
+ *
+ * <p>Example:
+ * <pre>{@code
+ * codestyleChecker {
+ *
+ *      javadocLinkChecker {
+ *
+ *          // Will report the broken FQN links if a number of them is exceed zero.
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: `if there is at least one.`

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 16]
```diff
@@ -28,21 +29,73 @@
...
+ *          threshold = 0
+ *
+ *          // Report consists of a logged violations.
+ *          reportType = "warn"
+ *      }
+ *
+ *      anotherCodestyleChecker {
+ *
+ *          // Will report the code style violations if a number of them is exceed five.
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: `if their number exceeds five.`

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 17]
```diff
@@ -28,21 +29,73 @@
...
+ *          threshold = 5
+ *
+ *          // Report consists of a logged violations and a failed build.
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: Remove `a` before `logged`.
It should be 
```
If threshold is exceeded, the build will fail. Its report will contain the violations.
```

File: spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java [hunk 18]
```diff
@@ -28,21 +29,73 @@
...
+ * <p>Example:
+ * <pre>{@code
+ * codestyleChecker {
+ *
+ *      javadocLinkChecker {
+ *
+ *          // Will report the broken FQN links if a number of them is exceed zero.
+ *          threshold = 0
+ *
+ *          // Report consists of a logged violations.
```
Comment (armiol) [inline spine-codestyle-checker/plugin/src/main/java/org/spine3/tools/codestyle/CodestyleCheckerPlugin.java]: `The violations are logged to the build report.`

**Pull request merged**